File: java-packages/src/main/java/com/rmaciak/order/domain/OrderRepository.java
package com.rmaciak.order.domain;

import org.springframework.stereotype.Repository;

import java.util.UUID;

import static com.rmaciak.order.domain.OrderEntity.from;

@Repository
class OrderRepository {

    private final OrderEntityPersistence orderPersistence;

    OrderRepository(OrderEntityPersistence orderPersistence) {
        this.orderPersistence = orderPersistence;
    }

    Order getOrder(UUID orderId) {
        return orderPersistence.getOrder(orderId).toDomain();
    }

    Order save(Order order) {
        return orderPersistence.save(from(order)).toDomain();
    }
}

File: kotlin-with-submodules/order-domain/src/main/kotlin/com/rmaciak/order/OrderRepository.kt
package com.rmaciak.order

import java.util.UUID

interface OrderPersistance {
    fun getOrderEntity(orderId: UUID): OrderEntity?
    fun saveOrder(order: OrderEntity): OrderEntity
}

class OrderRepository(
        private val orderPersistance: OrderPersistance
) {

    internal fun getOrder(orderId: UUID): Order? =
            orderPersistance.getOrderEntity(orderId)?.toDomain()

    internal fun saveOrder(order: Order): Order =
            orderPersistance.saveOrder(order.toEntity()).toDomain()
}

private fun OrderEntity.toDomain() =
        Order(
                id = this.id,
                quota = this.quota,
                items = this.items.map { it.toDomain() }.toSet()
        )

private fun OrderItemEntity.toDomain() =
        OrderItem(
                id = this.id,
                quota = this.quota
        )

